ai_spider/stats.py

```diff
@@ -6,30 +6,109 @@
 from threading import Thread
 
 from ai_spider.unique_queue import UniqueQueue
+from database.schema import StatsContainer as StatsContainerTable, StatsWorker as StatsWorkerTable, StatsBin as StatsBinTable
+from sqlalchemy import and_
 
 STATS_EMA_ALPHA = 0.9
 PUNISH_SECS = 60 * 15
 PUNISH_BAD_PERF = 9999
 
+class StatsStore:
+    def __init__(self, session=None):
+        self.update_queue = UniqueQueue(key=lambda el: el[0])
+        self.update_thread = Thread(daemon=True, target=self.update_loop)
+        self.update_thread.start()
+        self.session=session
+
+    def stop(self, join=True):
+        self.update_queue.put([None, None])
+        if join:
+            self.update_thread.join()
+
+    def update_loop(self):
+        while True:
+            try:
+                key, val = self.update_queue.get(timeout=4)
+                if key is None:
+                    break
+                self._update(key, val)
+            except Empty:
+                pass
+
+    def update(self, key: str, val: dict):
+        self.update_queue.put([key, val])
+
+    def _update(self, key, val):
+        self.session.add(val)
+        self.session.commit()
+        return val
+
+    def get(self,model: object, filters: dict) -> dict:
+        conditions = [getattr(model, column_name) == value for column_name, value in filters.items()]
+        rows = self.session.query(model).filter(and_(*conditions)).all()
+        return rows
+
 
 class StatsBin:
-    def __init__(self, alpha, val=None):
+    def __init__(self, alpha, row_id: int, msize: int, val=None, store: StatsStore = None):
+        unique_fields = {
+            "stats_worker_id": row_id,
+            "msize": msize
+        }
+        rows = store.get(StatsBinTable, unique_fields)
+        numRows = len(rows)
+        if numRows == 0:
+            dict = {
+                row_id: StatsBinTable(stats_worker_id=row_id, msize=msize, val=val)
+            }
+            self.row_id = store._update(key=row_id,val=dict)
+        elif numRows == 1:
+            firstRow = next(iter(rows))
+            self.row_id = rows[firstRow].id
+        else:
+            print(f"{numRows} stats_bin rows found but expected 0 or 1")
+            
         self.val = val
         self.alpha = alpha
+        self.store = store
 
     def bump(self, secs):
         if self.val is None:
             self.val = secs
         else:
             self.val = self.alpha * self.val + (1 - self.alpha) * secs
+        dict = {
+                secs: StatsBinTable(val=self.val)
+            }
+        self.store.update(key=self.row_id,val=dict)
+       
+        
 
 
 class StatsWorker:
-    def __init__(self, alpha):
+    def __init__(self, alpha, row_id: int,  store: StatsStore = None):
+        unique_fields = {
+            "stats_container_id": row_id,
+        }
+        rows = store.get(StatsWorkerTable,unique_fields)
```

Comment:
this is abstraction leaking.  the store should have knowledge of the table.  the caller should never know the table.

Reply:
maybe I have this backwards. If I keep the store generic I can do `get` calls like this. If I want this class to not have to know about the table, won't I need specific `getStatsWorker` table types of functions in the `StatsStore`

Reply:
the MySqlStatsStore derived specialization class should know about the table, connection, host, etc..